### PR TITLE
Cross platform powershell client.

### PR DESCRIPTION
**Pull request by george-giloan**

### CODE REVIEW

File: score-powershell/src/main/java/io/cloudslang/content/actions/PowerShellScriptAction.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            maxEnvelopeSize = StringUtils.isBlank(maxEnvelopeSize) ? InputDefaults.MAX_ENVELOPE_SIZE.getValue() : maxEnvelopeSize;
+            Integer.parseInt(maxEnvelopeSize);
+            trustAllRoots = StringUtils.isBlank(trustAllRoots) ? InputDefaults.TRUST_ALL_ROOTS.getValue() : trustAllRoots;
+            x509HostnameVerifier = StringUtils.isBlank(x509HostnameVerifier) ? InputDefaults.X_509_HOSTNAME_VERIFIER.getValue() : x509HostnameVerifier;
+            operationTimeout = StringUtils.isBlank(operationTimeout) ? InputDefaults.OPERATION_TIMEOUT.getValue() : operationTimeout;
+            winrmLocale = StringUtils.isBlank(winrmLocale) ? InputDefaults.WINRM_LOCALE.getValue() : winrmLocale;
+
+            resultMap = wsManRemoteShellService.runCommand(host, port, protocol, username, password, proxyHost,
+                    proxyPort, proxyUsername, proxyPassword, maxEnvelopeSize, trustAllRoots, x509HostnameVerifier, keystore, keystorePassword,
+                    trustKeystore, trustPassword, script, winrmLocale, Integer.parseInt(operationTimeout));
```
Comment (gituser173) [inline score-powershell/src/main/java/io/cloudslang/content/actions/PowerShellScriptAction.java]: maybe you could use an object to hold all these inputs, see score-vmware on how the builder pattern is used there for this purpose

File: score-powershell/src/main/java/io/cloudslang/content/services/WSManRemoteShellService.java [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+     * @param trustAllRoots
+     * @param x509HostnameVerifier
+     * @param keystore
+     * @param keystorePassword
+     * @param trustKeystore
+     * @param trustPassword
+     * @return the configured HttpClientInputs object.
+     * @throws MalformedURLException
+     */
+    public HttpClientInputs setCommonHttpInputs(HttpClientInputs httpClientInputs, URL url, String username, String password, String proxyHost, String proxyPort,
```
Comment (gituser173) [inline score-powershell/src/main/java/io/cloudslang/content/services/WSManRemoteShellService.java]: why is this method public?

File: score-powershell/src/main/java/io/cloudslang/content/actions/PowerShellScriptAction.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        try {
+            WSManRemoteShellService wsManRemoteShellService = new WSManRemoteShellService();
+            protocol = StringUtils.isBlank(protocol) ? InputDefaults.PROTOCOL.getValue() : protocol;
+            port = StringUtils.isBlank(port) ? InputDefaults.PORT.getValue() : port;
+            maxEnvelopeSize = StringUtils.isBlank(maxEnvelopeSize) ? InputDefaults.MAX_ENVELOPE_SIZE.getValue() : maxEnvelopeSize;
+            Integer.parseInt(maxEnvelopeSize);
+            trustAllRoots = StringUtils.isBlank(trustAllRoots) ? InputDefaults.TRUST_ALL_ROOTS.getValue() : trustAllRoots;
+            x509HostnameVerifier = StringUtils.isBlank(x509HostnameVerifier) ? InputDefaults.X_509_HOSTNAME_VERIFIER.getValue() : x509HostnameVerifier;
+            operationTimeout = StringUtils.isBlank(operationTimeout) ? InputDefaults.OPERATION_TIMEOUT.getValue() : operationTimeout;
+            winrmLocale = StringUtils.isBlank(winrmLocale) ? InputDefaults.WINRM_LOCALE.getValue() : winrmLocale;
```
Comment (gituser173) [inline score-powershell/src/main/java/io/cloudslang/content/actions/PowerShellScriptAction.java]: after you've created the inputs object maybe you could move this validation somewhere else, we can take this offline

File: score-powershell/src/main/java/io/cloudslang/content/services/WSManRemoteShellService.java [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+
+    /**
+     * This method executes a request with the given ScoreHttpClient, HttpClientInputs and body.
+     *
+     * @param scoreHttpClient
+     * @param httpClientInputs
+     * @param requestMessage
+     * @return the result of the request execution.
+     */
+    protected Map<String, String> executeRequest(ScoreHttpClient scoreHttpClient, HttpClientInputs httpClientInputs, String requestMessage) {
```
Comment (gituser173) [inline score-powershell/src/main/java/io/cloudslang/content/services/WSManRemoteShellService.java]: is this method protected for testing purposes?

File: score-powershell/src/main/java/io/cloudslang/content/services/WSManRemoteShellService.java [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+     * @throws SAXException
+     */
+    public Map<String, String> runCommand(WSManRequestInputs wsManRequestInputs) throws RuntimeException, IOException, InterruptedException, ParserConfigurationException, TransformerException, XPathExpressionException, TimeoutException, URISyntaxException, SAXException {
+        ScoreHttpClient scoreHttpClient = new ScoreHttpClient();
+        HttpClientInputs httpClientInputs = new HttpClientInputs();
+        URL url = buildURL(wsManRequestInputs, WSMAN_RESOURCE_URI);
+        httpClientInputs = setCommonHttpInputs(httpClientInputs, url, wsManRequestInputs);
+        String shellId = createShell(scoreHttpClient, httpClientInputs, CREATE_SHELL_REQUEST_XML, url.toString(), wsManRequestInputs);
+        WSManUtils.validateUUID(shellId, SHELL_ID);
+        String commandStr = "PowerShell" + " " + "-NonInteractive" + " " + "-EncodedCommand" + " " + EncoderDecoder.encodeStringInBase64(wsManRequestInputs.getScript(), Charsets.UTF_16LE);
```
Comment (gituser173) [inline score-powershell/src/main/java/io/cloudslang/content/services/WSManRemoteShellService.java]: maybe extract constant here COMMAND_PREFIX?

File: score-powershell/src/main/java/io/cloudslang/content/services/WSManRemoteShellService.java [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+     * @throws SAXException
+     * @throws ParserConfigurationException
+     */
+    private String createShell(ScoreHttpClient scoreHttpClient, HttpClientInputs httpClientInputs, String resourceName, String url,
+                               WSManRequestInputs wsManRequestInputs) throws RuntimeException, IOException, URISyntaxException,
+            TransformerException, XPathExpressionException, SAXException, ParserConfigurationException {
+        String document = ResourceLoader.loadAsString(resourceName);
+        document = createCreateShellRequestBody(document, url, String.valueOf(wsManRequestInputs.getMaxEnvelopeSize()),
+                wsManRequestInputs.getWinrmLocale(), String.valueOf(wsManRequestInputs.getOperationTimeout()));
+        Map<String, String> createShellResult = executeRequest(scoreHttpClient, httpClientInputs, document);
```
Comment (gituser173) [inline score-powershell/src/main/java/io/cloudslang/content/services/WSManRemoteShellService.java]: the code below this comment seems to be duplicated in the executeCommand method, maybe you could extract it in a new method

**Comment by gituser173** [comment]
please handle the comments, also more unit tests should be added.

**Comment by george-giloan** [comment]
Comments have been handled @gituser173

**Pull request merged**